### DISCUSSION TITLE
Case Sensitivity

### DISCUSSION THREAD

**Discussion by dallasbeek** [CoreWCF/CoreWCF#1346] [General]
I have an operation defined as
    [OperationContract]
    [WebInvoke(Method = "POST", BodyStyle = WebMessageBodyStyle.Wrapped,
        RequestFormat = WebMessageFormat.Json, ResponseFormat = WebMessageFormat.Json)]
    Task<UploadResult> SubmitData(string authToken);

But if I post data { "authtoken": "ABC" } my request does not receive the data as the authToken parameter appear to be case sensitive.  Is there a way to make arguments case insensitive?

**Comment by mconnew** [comment]
Internally, CoreWCF uses DataContractJsonSerializer to deserialize the parameter. For each parameter on the method, it creates a separate instance so that each parameter is individually deserialized. We have a list of part for each operation, and the part is what holds the serializer. When an incoming message arrives, we consume the outer JSON object wrapper which forms the entire body, ie the starting `{` of the body. Parameters get expressed as JSON object properties.  

Because a parameter might be omitted, when we read the object properties, we can't presume it's the next item. E.g. if your method was defined as `Task DoSomething(string foo, string bar)` and we received the JSON `{ "bar" : "my bar string values" }`, we need to check the name of the `"bar"` property against each of the defined parts so that we match it up with the `bar` parameter instead of the `foo` parameter. The way we validate this is to call `DataContractJsonSerializer.IsStartObject(XmlDictionaryReader reader)`. I know it seems odd to pass an `XmlDictionaryReader` as this is JSON, but we have an implementation which actually reads JSON and presents it as though it's XML. JSON has a subset of features that XML has so we can pretend it's an XML document and with things like attributes, the JSON XmlReader acts as though there are no attributes.  

`DataContractJsonSerializer.IsStartObject` checks whether the XmlReader is positioned at the start of an "element" (in xml this would be `<authToken>`, but for JSON it means we're positioned at the "authtoken" property name) that the serializer is able to deserialize from. For my `DoSomething` method from earlier, the serializer for the `foo` parameter will return `false` when the reader is positioned at `bar`, but the serializer for `bar` will return true.  

I thought this would be easy for you to solve yourself by replacing the DataContractJsonSerializerOperationBehavior/DataContractSerializerOperationBehavior on your operation and providing a wrapped serializer which is changes the behavior of `IsStartObject` to be case insensitive, but WebHttpBehavior steps outside of the normal way of doing things and instantiates its own instance and only takes property values from an operation attached behavior. It turns out doing this is a lot of work, beyond what I would expect to be reasonable for most people to do. So I coded it up for you. I've attached to this comment the full implementation along with a sample project which does this. 

It's opt in on a per-operation basis to try to avoid the risk of any behavior changes between this and the in-box implementation. You modify `CaseInsensitiveDeserializeEndpointBehavior.ShouldReplaceFormatter` to select the operations you want this implementation to be used for. It only replaces the incoming deserializing logic and leaves the serializing reply logic as the original implementation. This can't switch between Xml, Json and Raw, it only supports Json. You could make it switch by replacing the `DeserializeRequest` method in `CaseInsensitiveDeserializeDispatchFormatter` (you need to use the `new` modifier on the method as the base method isn't virtual), detecting it's not Json and calling `DeserializeRequest` on the inner implementation. This was enough code to put together as it is and without knowing you need that, I omitted it.  

To use this, copy CaseInsensitiveDeserializeServiceBehavior.cs into your project and add a singleton of type `IServiceBehavior` to your services collection. Something like:
```c#
builder.Services.AddSingleton<IServiceBehavior, CaseInsensitiveDeserializeServiceBehavior>();
```
[GenericWebHttpCoreWCFService.zip](https://github.com/CoreWCF/CoreWCF/files/14500539/GenericWebHttpCoreWCFService.zip)

**Comment by dallasbeek** [comment]
Thank you for this, I am reviewing.  I will update you with my findings.

**Comment by Noah1989** [comment]
The example works in general. Here are some things I found out while using it:

- It only handles the wrapped body style out of the box, implementing the non-wrapped case is left as an exercise for the reader.
- The reply handling is broken (it always returns XML). That is because shadowing the SerializeReply method with `new` is not effective. You have to instead make en explicit interface implementation of `IDispatchMessageFormatter.SerializeReply`, then it is actually being called.
